src/hooks/useLanguage/useLanguage.ts

```diff
@@ -0,0 +1,80 @@
+import { useCallback, useEffect, useState } from "react";
+import { useLocation, useHistory } from "react-router-dom";
+import { Language, Languages, LANGUAGES } from "../../constants/languages";
+import { useQueryParams } from "../../hooks/useQueryParams";
+
+const defaultLang: Language = Languages.TypeScript;
+
+const LOCAL_KEY = "preferred-language";
+const PARAM_KEY = "lang";
+
+const isValidLang = (lang: Language | void) => lang && LANGUAGES.includes(lang);
+
+const getInitialLang = (langFromParams: Language): Language => {
+  // First, use language from query params in url
+  if (isValidLang(langFromParams)) {
+    return langFromParams as Language;
+  }
+
+  // Next check for one stored in localStorage
+  try {
+    const storedLang = (localStorage.getItem(LOCAL_KEY) ?? "") as Language;
+    if (isValidLang(storedLang)) return storedLang;
+  } catch {
+    // Do nothing, we just don't want to crash if localStorage access is blocked.
+  }
+
+  // Otherwise fallback to a default
+  return defaultLang;
+};
+
+export interface UseLanguageOptions {
+  /**
+   * Syncs the preferred language to a query param in URL
+   */
+  updateUrl?: boolean;
+  /**
+   * Saves the selected language to localStorage on select
+   */
+  updateSaved?: boolean;
+}
```

Comment:
This was changed to a hook rather than context per @MrArnoldPalmer's suggestion as all the info it needs regarding initial state exists within the URL or localStorage.

The usecases for these options are the following:

updateUrl - package page will always have lang in URL (meaning sharing links will provide a consistent experience)

updateSaved - eventually we will want to create a user preference page where a user can set a preferred theme. As such, we'll no longer want to overwrite that on the package page